addons/l10n_in_edi/models/account_move.py

```diff
@@ -61,3 +62,17 @@ def _l10n_in_edi_is_managing_invoice_negative_lines_allowed(self):
         """
         param_name = 'l10n_in_edi.manage_invoice_negative_lines'
         return bool(self.env['ir.config_parameter'].sudo().get_param(param_name))
+
+    @api.depends('move_type', 'line_ids.tax_tag_ids', 'amount_residual', 'partner_id.l10n_in_pan')
+    def _compute_l10n_in_edi_display_null_pan_warning(self):
+        for record in self:
+            if record.country_code == 'IN':
+                tds_tcs_tax_tags = record.env.ref("l10n_in.tax_tag_base_tcs") | record.env.ref("l10n_in.tax_tag_base_tds")
+                record.display_null_pan_warning = (
+                    not record.partner_id.l10n_in_pan
+                    and record.move_type in ['in_invoice', 'out_invoice']
+                    and record.amount_residual > 0
+                    and record.line_ids.mapped("tax_tag_ids") & tds_tcs_tax_tags
+                )
+            else:
+                record.display_null_pan_warning = False
```

Comment:
```suggestion
                record.l10n_in_display_null_pan_warning = (
                    record.country_code == 'IN'
                    and not record.partner_id.l10n_in_pan
                    and record.move_type in ['in_invoice', 'out_invoice']
                    and record.amount_residual > 0
                    and (
                        record.line_ids.mapped("tax_tag_ids")
                        & (self.env.ref("l10n_in.tax_tag_base_tcs") + self.env.ref("l10n_in.tax_tag_base_tds"))
                    )
                )
```
@jasa-odoo You can simply use short-cuiting rather than using if-else since it is a boolean field. Well though up to because readability is as much important too